Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -73,132 +100,267 @@
 %
-% For model fitting, we want the PLDs-first-TEs-second order (just like the
-% beginning) so we need to reorder it again
-if isfield(xQ,'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
+% And for decoding we want to swap the dimensions such that the Hadamard blocks (==PLDs) are the innermost dimension:
+% PLD1 TE1 | PLD2 TE1 | PLD3 TE1 | PLD4 TE1 | PLD1 TE2 | PLD2 TE2 | PLD3 TE2 | PLD4 TE2 | ... | PLD1 TE8 | PLD2-TE8 | PLD3 TE8 | PLD4-TE8
+% (PLDs in the first dimension, TEs after)
+
+imEncodedReordered = imEncoded;
+
+if nInnerRepetitions > 1 % Reordering is only needed when having inner repetitions
 	% This is the original order
-	vectorOldOrder = 1:size(imDecoded, 4);
-	
-	% Shape to a matrix with TEs second and all the rest first
-	vectorOldOrder = reshape(vectorOldOrder, size(imDecoded, 4)/xQ.NumberEchoTimes, xQ.NumberEchoTimes);
-	
-	% Flip the two dimensions and make a row vector again
-	vectorOldOrder = reshape(vectorOldOrder', 1, size(imDecoded, 4));
+	vectorReorderEncoded = 1:nEncodedVolumes;
+
+	% Shape to a matrix with TEs first and all the rest later
+	vectorReorderEncoded = reshape(vectorReorderEncoded, nInnerRepetitions, nEncodedVolumes/nInnerRepetitions);
+
+	% Switch the two dimensions and create a row vector again
+	vectorReorderEncoded = reshape(vectorReorderEncoded', 1, nEncodedVolumes);
 
 	% Reorder the data
-	imDecoded = imDecoded(:,:,:,vectorOldOrder);
+	imEncodedReordered = imEncodedReordered(:,:,:,vectorReorderEncoded);
+end
+% Sometimes an acquisition (e.g., Hadamard-8) is: outerRepetition PLD -> PLD1 rep1 | PLD1 rep2 | ... PLD2 rep1 | PLD2 rep2
+% We want here                                    PLD outerRepetition -> PLD1 rep1 | PLD2 rep1 | ... PLD1 rep2 | PLD2 rep2
+
+
+%% 6. Decode the image volumes
+% We assume that the all encoded volumes are ordered as repetitions of Hadamard blocks
+
+nPLDs = nBlocks-1;
+% We only decode the PLD blocks here, we ignore the control volumes
+% Hence the number of PLDs is the total number of Hadamard blocks (TimeEncodedMatrixSize - 1)
+
+decodedPWI4D = zeros(size(imEncodedReordered,1), size(imEncodedReordered,2), size(imEncodedReordered,3), nOuterRepetitions * nPLDs);
+
+% Go through all repetitions of the full Hadamard blocks
+for iRepetition = 1:nOuterRepetitions
+	offsetEncoded = nBlocks * (iRepetition-1); % Index offset for the given repetition to search in the encoded data
+	encodedBlocks = imEncodedReordered(:,:,:,(1:nBlocks) + offsetEncoded); % Obtain the single encoded block
+	decodedBlocks = zeros(size(imEncodedReordered,1), size(imEncodedReordered,2), size(imEncodedReordered,3), nPLDs); % Temporary single decoded Hadamard block
+
+	for iBlock = 1:nPLDs
+
+		indexPositive = find(xQ.TimeEncodedMatrix(iBlock,:)==1);
+		indexNegative = find(xQ.TimeEncodedMatrix(iBlock,:)==-1);
+		decodedBlocks(:,:,:,iBlock) = sum(encodedBlocks(:,:,:,(indexPositive)),4) - sum(encodedBlocks(:,:,:,(indexNegative)),4);
```

Comment:
If we have resampled volumes, e.g., because of motion correction, there is the possibility for having nonfinite voxels (e.g., NaN) for some of the volumes. This will be mostly outside of the original FoV but you never know what happens :)

Do you want to have a check for nonfinite values or do you want to replace sum here by xASL_stat_SumNan? I have no preference.

Reply:
If, for a given voxel, one of the values is NaN, then this is completely messed up. The innerRepetitions are not equal and we have to throw away the entire voxel. Of course, for outerRepetitions, we can take only non-NaN part and work with it. But makes no sense to use sumNan for the innerRepetitions

Reply:
Here you are not working at innerRepetitions, you are summing/decoding Hadamard blocks/volumes.
If you say that a single NaN will completely mess it up, then we can keep it like this.
It would be nice to add a warning, but then we would need to check that we only check voxels inside the brainmask for this warning, because we don't want to issue a warning for NaN-voxels outside the brain/original FoV (because of motion resampling), because that is normal to occur.

So conclusion: we keep it like it is, OK? If you agree please resolve this conversation.

Reply:
yes. sorry. this is what I meant. If one hadamard volume is bad, then we have to throw away the complete stuff for the specific voxel. For multi-PLD, we can throw away a single PLD and fit the rest. For control-label, we can throw one control-label pair out of many. Or we can also throw a single control out and create the average control of the rest. For hadamard, if a single volume is out, we can't reconstruct the PLDs from the rest -> we really need all 4 volumes to properly reconstruct all 3 PLDs. If a single one is missing, it won't work. We would have to take a similar outerRepetition to replace for it.

So I guess we have to keep the code like it is.